Add distributed RDataFrame pyspark build option

vepadulano opened the pull request.

vepadulano commented:
@phsft-bot build with flags -Ddataframe_distributed_pyspark=ON

vepadulano commented:
@phsft-bot build with flags -Ddataframe_distributed_pyspark=ON

oshadura commented on cmake/modules/RootBuildOptions.cmake:
```diff
@@ -117,6 +117,7 @@ ROOT_BUILD_OPTION(cuda OFF "Enable support for CUDA (requires CUDA toolkit >= 7.
 ROOT_BUILD_OPTION(cudnn ON "Enable support for cuDNN (default when Cuda is enabled)")
 ROOT_BUILD_OPTION(cxxmodules OFF "Enable support for C++ modules")
 ROOT_BUILD_OPTION(dataframe ON "Enable ROOT RDataFrame")
+ROOT_BUILD_OPTION(dataframe_distributed_pyspark OFF "Enable distributed RDataFrame pyspark python module")
```
I think it could be a bit long as a build option? (I imagine I will not be happy to type it. :)

oshadura replied:
``` dataframe_distpyspark``?

vepadulano replied:
Sounds good, thanks!

vepadulano commented:
Latest commit addresses suggestions (shortened build option name) and improves on the logic to find pyspark and its dependencies. This is ready for review now.

Axel-Naumann commented on cmake/modules/FindPySpark.cmake:
```diff
@@ -0,0 +1,74 @@
+
+    if(${_PY4J_SEARCH_SUCCESS} EQUAL 0)
+        message(STATUS "Package 'py4j' has been succesfully imported with ${PYTHON_EXECUTABLE_Development_Main}")
+        message(STATUS "Import python package: pyspark")
```
```suggestion
        message(STATUS "Looking for python package: pyspark")
```
("Import python package: pyspark" sounds like a directive to the user, or to me it's less clear what it means)

vepadulano replied:
Corrected, thanks!

Axel-Naumann commented on cmake/modules/FindPySpark.cmake:
```diff
@@ -0,0 +1,74 @@
+    message(STATUS "Java version ${Java_VERSION_STRING}")
+    message(STATUS "Looking for pyspark dependency: py4j")
+
+    execute_process(
+        COMMAND ${PYTHON_EXECUTABLE_Development_Main} -c "import py4j"
+        RESULT_VARIABLE _PY4J_SEARCH_SUCCESS
+        OUTPUT_VARIABLE _PY4J_VALUES_OUTPUT
+        ERROR_VARIABLE _PY4J_ERROR_VALUE
+        OUTPUT_STRIP_TRAILING_WHITESPACE
+    )
+
+    if(${_PY4J_SEARCH_SUCCESS} EQUAL 0)
```
What if we just skip this, and only try `import pyspark`? If that shows the error message ("cannot import py4j") it should be enough?

vepadulano replied:
Sounds good, done

Axel-Naumann commented on cmake/modules/FindPySpark.cmake:
```diff
@@ -0,0 +1,74 @@
+#  PYSPARK_FOUND - system has pyspark and its dependencies are met
+#
+
+message(STATUS "Looking for pyspark dependency: Java")
```
Shouldn't this define the function `findPySpark` (or whatever the CMake convention is)?

vepadulano replied:
I read a bit more about this since I'm not expert and Oksana gave me some very helpful suggestions. The CMake convention is to have a `FindX.cmake` file define a set of variables that flag the availability of the package in the system (for example the home directory of the package) and then use those as arguments to [find_package_handle_standard_args](https://cmake.org/cmake/help/v3.9/module/FindPackageHandleStandardArgs.html) . Examples we use in ROOT with python modules are in [FindNumpy.cmake](https://github.com/root-project/root/blob/a56901942dfefa6e7e2e7df397a60c9d1cda31a0/cmake/modules/FindNumPy.cmake#L100-L103) and the function [find_python_module](https://github.com/root-project/root/blob/a56901942dfefa6e7e2e7df397a60c9d1cda31a0/cmake/modules/RootMacros.cmake#L1908-L1909)

Axel-Naumann commented on cmake/modules/FindPySpark.cmake:
```diff
@@ -0,0 +1,74 @@
+#  PYSPARK_FOUND - system has pyspark and its dependencies are met
+#
+
+message(STATUS "Looking for pyspark dependency: Java")
```
Traditionally, `findXYZ` handle a `REQUIRED` parameter that errors out of `REQUIRED` but not found. This could be set for handling `fail-on-missing`, as done for the other `findPackage` cases.

vepadulano replied:
In the next commit `find_package(Java)` will have a `REQUIRED` parameter in all cases, since it's a strict requirement for pyspark that cannot be catched in `import pyspark` unfortunately

Axel-Naumann commented on cmake/modules/FindPySpark.cmake:
```diff
@@ -0,0 +1,74 @@
+#
+
+message(STATUS "Looking for pyspark dependency: Java")
+find_package(Java 1.8 COMPONENTS Runtime)
```
Would `import pyspark` work without java runtime installed? If yes: I agree, we need to check for the Java runtime. If it would fail, let's just rely on `pyspark` checking it?

vepadulano replied:
Unfortunately `import pyspark` doesn't error out if Java runtime is not installed on the system. It only errors out during runtime of a pyspark application

Axel-Naumann commented on cmake/modules/FindPySpark.cmake:
```diff
@@ -0,0 +1,74 @@
+            if(${PYTHON_VERSION_STRING_Development_Main} VERSION_LESS "3.6")
+                # Pyspark dropped support for Python < 3.6 in version 3.1.1
+                if(${_PYSPARK_VERSION_STRING} VERSION_GREATER_EQUAL "3.1.1")
+                    message(STATUS "Package 'pyspark' found with version ${_PYSPARK_VERSION_STRING}, which requires Python > 3.5")
```
Yet the `import` would nonetheless work? That sounds like a bug in pyspark: if a Python version isn't supported, I'd expect the `import` to fail!

vepadulano replied:
Indeed this should not be needed, will be removed in the next commit

Axel-Naumann commented on cmake/modules/FindPySpark.cmake:
```diff
@@ -0,0 +1,74 @@
+                    message(STATUS "Please install a version of 'pyspark' compatible with your Python executable")
+                    set(PYSPARK_FOUND FALSE)
+                else()
+                    message(STATUS "Package 'pyspark' has been succesfully imported with ${PYTHON_EXECUTABLE_Development_Main}")
```
```suggestion
                    message(STATUS "Package 'pyspark' version ${_PYSPARK_VERSION_STRING} has been successfully imported with ${PYTHON_EXECUTABLE_Development_Main}")
```

vepadulano replied:
In the next commit the pyspark version string will be automatically handled through `find_package_handle_standard_args` so the output will be more in line with common CMake logic

Axel-Naumann commented on cmake/modules/FindPySpark.cmake:
```diff
@@ -0,0 +1,74 @@
+                    set(PYSPARK_FOUND TRUE)
+                endif()
+            else()
+                message(STATUS "Package 'pyspark' has been succesfully imported with ${PYTHON_EXECUTABLE_Development_Main}")
```
See above

Axel-Naumann commented on cmake/modules/RootBuildOptions.cmake:
```diff
@@ -482,3 +484,12 @@ elseif(Python3_Interpreter_Development_FOUND AND Python2_Interpreter_Development
+  message(STATUS "    Switching it OFF because either pyroot or dataframe option is disabled")
+  set(dataframe_distpyspark OFF CACHE BOOL "Disabled because either dataframe or pyroot were disabled" FORCE)
+
+endif()
```
Could you add a newline at the end of the file, please?

Axel-Naumann commented on cmake/modules/SearchInstalledSoftware.cmake:
```diff
@@ -1865,3 +1865,31 @@ if(NOT ROOT_HAVE_CXX_ATOMICS_WITHOUT_LIB)
+
+#------------------------------------------------------------------------------------
+# Check if the pyspark package is installed on the system.
+# The FindPySpark.cmake module will first look for pyspark dependencies:
+# 1. It will look for Java in the system. Minimum version 1.8
+# 2. It will try to `import py4j` using PYTHON_EXECUTABLE_Development_Main
```
I don't think this needs to be called out here; to me that's an implementation detail of `FindPySpark.cmake`.

Axel-Naumann commented on cmake/modules/SearchInstalledSoftware.cmake:
```diff
@@ -1865,3 +1865,31 @@ if(NOT ROOT_HAVE_CXX_ATOMICS_WITHOUT_LIB)
+# 1. It will look for Java in the system. Minimum version 1.8
+# 2. It will try to `import py4j` using PYTHON_EXECUTABLE_Development_Main
+#
+# If the two depencies are met, the module will try to `import pyspark` using the same
+# Python executable. If everything could be found then the PYSPARK_FOUND variable will
```
Maybe "is usable" instead of spelling out the (current) dependencies of pyspark?
```suggestion
# If `pyspark` can be imported and is usable then the PYSPARK_FOUND variable will
```

Axel-Naumann commented on cmake/modules/FindPySpark.cmake:
```diff
@@ -0,0 +1,74 @@
+        set(PYSPARK_FOUND FALSE)
+    endif()
+else()
+    set(PYSPARK_FOUND FALSE)
```
Shouldn't `PYSPARK_FOUND` be cached?

vepadulano replied:
Following Oksana's suggestion, the [find_package_handle_standard_args](https://cmake.org/cmake/help/v3.9/module/FindPackageHandleStandardArgs.html) function is a better way to signal CMake that pyspark has been found correctly. It will internally set the `PySpark_FOUND` variable

Axel-Naumann reviewed (commented):
We need @oshadura here!

oshadura commented on cmake/modules/FindPySpark.cmake:
```diff
@@ -0,0 +1,74 @@
+            OUTPUT_STRIP_TRAILING_WHITESPACE
+        )
+
+        string(REGEX MATCH "^[0-9]+\\.[0-9]+\\.[0-9]+" _PYSPARK_VERSION_STRING "${_PYSPARK_VALUES_OUTPUT}")
```
Let's maybe save _PYSPARK_VERSION_STRING? It could be useful.

vepadulano replied:
Yes! The new commit will do this

oshadura commented on cmake/modules/FindPySpark.cmake:
```diff
@@ -0,0 +1,74 @@
+        message(STATUS "Import python package: pyspark")
+
+        execute_process(
+            COMMAND ${PYTHON_EXECUTABLE_Development_Main} -c "import pyspark; print(pyspark.__version__)"
```
we are checking pyspark only for PYTHON_EXECUTABLE_Development_Main, right? We need to mention it somewhere (e.g. README or notes)...

vepadulano replied:
It will be mentioned in the FindPySpark module and also in the relnotes :+1:

oshadura commented on cmake/modules/FindPySpark.cmake:
```diff
@@ -0,0 +1,74 @@
+else()
+    set(PYSPARK_FOUND FALSE)
+endif()
+
```
I am thinking the logic of checks for PYSPARK_FOUND is a bit complex...
It will be nice if we could handle PySpark_FOUND here in more 'traditional way': 
```
include(FindPackageHandleStandardArgs)
find_package_handle_standard_args(PySpark
  REQUIRED_VARS PYSPARK_VERSION ..
)
```
then you will get **PySpark_FOUND** automatically.

vepadulano replied:
The new commit will simplify the logic

oshadura commented on cmake/modules/FindPySpark.cmake:
```diff
@@ -0,0 +1,74 @@
+    message(STATUS "Java version ${Java_VERSION_STRING}")
+    message(STATUS "Looking for pyspark dependency: py4j")
+
+    execute_process(
```
Check `find_python_module()` code, maybe it will be useful to sync it here: https://github.com/root-project/root/blob/master/cmake/modules/RootMacros.cmake#L1878

vepadulano replied:
Definitely, thanks!

oshadura commented on cmake/modules/FindPySpark.cmake:
```diff
@@ -0,0 +1,74 @@
+#
+
+message(STATUS "Looking for pyspark dependency: Java")
+find_package(Java 1.8 COMPONENTS Runtime)
```
Did you test distributed RDataframe + pyspark with both Java 1.8 and 1.9 >? (just in case checking)

vepadulano replied:
It has been tested with Java 1.8 and Java 11, which are the default respectively for Spark 2.4 (LTS of Spark 2.x) and Spark 3.x

vepadulano commented:
@phsft-bot build just on mac1014, ROOT-fedora30, ROOT-ubuntu2004 with flags -Ddataframe_distpyspark=ON

vepadulano commented:
@phsft-bot build just on mac1014/default, ROOT-fedora30/default, ROOT-ubuntu2004/default with flags -Ddataframe_distpyspark=ON

vepadulano commented:
@phsft-bot build also on mac1014/default, ROOT-fedora30/default, ROOT-ubuntu2004/default with flags -Ddataframe_distpyspark=ON

vepadulano commented:
@phsft-bot build just on mac1014/default, ROOT-fedora30/default, ROOT-ubuntu2004/default with flags -Ddataframe_distpyspark=ON

vepadulano commented:
Latest commit addresses all comments. Examples from the build nodes:
Failure example: [ROOT-performance-centos8-multicore/default](https://lcgapp-services.cern.ch/root-jenkins/job/root-pullrequests-build/109377/console) (pyspark is not installed on this system)
```
13:04:01  -- Looking for PySpark
13:04:01  -- Looking for PySpark dependency: Java
13:04:01  -- Found Java: /usr/bin/java (found suitable version "1.8.0.275", minimum required is "1.8") found components:  Runtime 
13:04:01  -- Found Java /usr/bin/java
13:04:01  -- Java version 1.8.0_275
13:04:01  -- Python package 'pyspark' could not be imported with /usr/bin/python
13:04:01  Traceback (most recent call last):
13:04:01    File "<string>", line 1, in <module>
13:04:01  ModuleNotFoundError: No module named 'pyspark'
13:04:01  
13:04:01  CMake Error at /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
13:04:01    Could NOT find PySpark (missing: PySpark_HOME) (Required is at least
13:04:01    version "2.4")
13:04:01  Call Stack (most recent call first):
13:04:01    /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
13:04:01    cmake/modules/FindPySpark.cmake:55 (find_package_handle_standard_args)
13:04:01    cmake/modules/SearchInstalledSoftware.cmake:16 (_find_package)
13:04:01    cmake/modules/SearchInstalledSoftware.cmake:1876 (find_package)
13:04:01    CMakeLists.txt:219 (include)
```

Success example: [ROOT-fedora30/default](https://lcgapp-services.cern.ch/root-jenkins/job/root-pullrequests-build/109385/consoleFull) 
```
13:05:04  -- Looking for PySpark
13:05:04  -- Looking for PySpark dependency: Java
13:05:04  -- Found Java: /usr/bin/java (found suitable version "1.8.0.252", minimum required is "1.8") found components: Runtime 
13:05:04  -- Found Java /usr/bin/java
13:05:04  -- Java version 1.8.0_252
13:05:04  -- Found PySpark: /usr/local/lib/python3.7/site-packages/pyspark (found suitable version "3.0.1", minimum required is "2.4") 
```

vepadulano commented:
Forgot to update the copyright in FindPySpark.cmake

vepadulano commented:
@phsft-bot build just on mac11.0/default, ROOT-fedora32/default, ROOT-ubuntu2004/default with flags -Ddataframe_distpyspark=ON
